### ISSUE TITLE
Support for Mayflash DolphinBar (braindump + RFC)

### ISSUE THREAD

**Issue by janoc** [rpavlik/wiiuse#66]
The DolphinBar is an interesting device. For $20 you can get a "sensor bar" (http://www.mayflash.com/Products/NINTENDOWiiGC/W010.html) that eliminates essentially all difficulties in using Wiimotes on a PC, including the latest -TR ones.

It works as a sensor bar (i.e. provides the tracking IR LEDs) but also acts as a translator between the Nintendo version of the HID over Bluetooth and normal USB HID. This makes pairing Wiimotes effortless (they pair to the bar), the PC support is also very simple, because the device acts as 4 HID controllers (gamepads). There are no drivers to install. Especially for Windows users this makes the Wiimote usable without having to battle different Bluetooth stacks, deal with incompatible Bluetooth dongles, add/remove device drivers or deal with external utilities that break randomly. 

The bar has 4 modes, three focus on emulating mouse+keyboard or a gamepad, the mode 4 acts as a pass-through for the connected Wiimotes, re-sending the received Bluetooth HID messages as normal USB HID. 

The bar works with Wiiuse out of the box in Windows in mode 4, however it doesn't with Linux & (most likely) Mac. The reason is that the Windows code actually enumerates HID devices and thus will detect the 4 emulated HID devices from the bar as well. Linux & Mac code attempt to use Bluetooth directly and are launching discovery - this will obviously fail, because the Wiimote is connected to the bar already. 

If we want to support this device in Wiiuse, we need to look at how to integrate the platform-specific parts of HID better. E.g. Mac apparently allows to pair and connect a Wiimote similarly as Windows does (#65) and there is also similar support in recent Linux kernels. 

Implementation-wise we have two options:
1. Either we switch to the HID support, similar to how the Windows code works, breaking compatibility with older code, 
2. Either we implement the HID support as a separate "mode", allowing the client code to talk to the Wiimote either over Bluetooth directly or to go through HID. 

The HID support could probably  be handled by HIDAPI, saving us a lot of work dealing with platform-specific code.

**Comment by PureTryOut** [comment]
I've been using a DolphinBar in combination with the Dolphin emulator and it works amazingly. I'd love it to be supported under WiiUse as well as it's my only way of providing IR lights to my Wiimotes and thus can't use my Wiimotes as pointer devices atm. Besides it fixes all connection issues I'm having with unofficial and gen20 devices.

@janoc is WiiUse 2.0 still happening? I'm willing to buy a DolphinBar for any (proven) developer that wants to implement support for it. Looking at the git log that would be either you or @rpavlik.

**Comment by janoc** [comment]
Hello, 

There hasn't been much work on "Wiiuse 2.0" since that discussion, as there really wasn't much pressing need. Ryan has moved on to different projects and I am also not involved much with virtual reality anymore. However, I am trying to maintain Wiiuse in a working shape.  

The DolphinBar mentioned in 2015 was bought for my then employer, so I don't have access to it anymore. I am surprised that this device is still sold, 7 years later?

Supporting it within Wiiuse would be possible (for Windows it just works, AFAIK) but it would require some significant effort. The "Wiiuse 2.0" was brought up because the way Wiiuse connects to Wiimotes is rather "hacky" and non-standard Bluetooth. It was required at the time because there was nothing else, but these days Linux + Bluez support Wiimotes, including pairing (which Wiiuse never supported). However, grafting that functionality onto Wiiuse would be non-trivial, as that has a very different design philosophy. 

I am not sure I have the resources to take that on myself (plus I don't have really any use case for it nor the hw itself). However, if someone would like to contribute some work here and lead the effort, I will be happy to collaborate. 

Regards,

Jan

**Comment by PureTryOut** [comment]
> I am surprised that this device is still sold, 7 years later?

Yup. It's actually my 3rd. The first one suffered from a burned-out led but they have since fixed that issue with a firmware update. I somehow lost my 2nd one while moving things around in my house, and I bought a 3rd one early last year. You can get them through Amazon from the original seller still, and imo it's _the best_ way to connect any Wii peripherals to a PC. It even allows connecting buggy 3rdparty remotes which so far no other method on PC has managed to do for me.

>  but these days Linux + Bluez support Wiimotes, including pairing (which Wiiuse never supported).

Are you referring to the `hid-wiimote` module in the kernel itself? [XWiimote](https://github.com/dvdhrm/xwiimote) uses that and although it sounds promising, the fact that you _require_ to connect to your Wiimote _from_ the PC makes it a downgrade compared to WiiUse in my opinion. I want to be able to just press 1 and 2 at the same time and connect to any device that listens for it, just like the original Wii does. WiiUse allows this, XWiimote with hid-wiimote doesn't.

> plus I don't have really any use case for it nor the hw itself

The use-case I can't help you with, but I _can_ buy it for you. I don't have the skills or knowledge to add this support myself but am happy to provide support and testing where needed, as well as buying the bar for anyone that can help (with a limit of course :sweat_smile: ).

**Comment by janoc** [comment]
> Are you referring to the `hid-wiimote` module in the kernel itself? [XWiimote](https://github.com/dvdhrm/xwiimote) uses that and although it sounds promising, the fact that you _require_ to connect to your Wiimote _from_ the PC makes it a downgrade compared to WiiUse in my opinion. I want to be able to just press 1 and 2 at the same time and connect to any device that listens for it, just like the original Wii does. WiiUse allows this, XWiimote with hid-wiimote doesn't.

I don't remember the name of the module as it has been a while since I have looked at it. However, I do remember that once the Wiimote was paired with the PC, all you had to do was to turn it on and it would automatically connect to the PC, the same as it works with the Wii. The Wiimote is even available as a normal gamepad/joystick to the OS (OK, you can't do much with it) and you can see and manage it e.g. from KDE/Plasma. That's a clear upgrade in my book. 

Pressing 1+2 each time is actually a hack and doesn't even work with the newer (-TR) Wiimotes. There you need to press the button in the battery compartment (pressing 1+2 only wakes the Wiimote up). What this does is that it puts the Wiimote into discoverable/pairing mode and then Wiiuse can find it. However, the normal operation is that you do this once, pair the Wiimote and then it reconnects to whatever it was paired with last time automatically, without any user's intervention next time. 

This works with Wiiuse in Windows if you use some tool to pair the Wiimote (or the DolphinBar), because there Wiiuse doesn't do anything with Bluetooth. In Linux it is possible to make it work but the Wiiuse code doesn't work with it - it expects to always discover an advertising (discoverable) Wiimote and then connect to it.

**Comment by PureTryOut** [comment]
Interesting. Anyways, that's offtopic for the original issue: DolphinBar support. Like I said I'm willing to buy one for you and @rpavlik if it helps getting it supported, please let me know if you're interested.

**Comment by janoc** [comment]
I don't really have the means to commit to a big project like that atm. 

I can have a look once again at what it would take to support HID in Linux (because that's what we need for DolphinBar as well) but I am afraid that would require supporting things like D-BUS - device connection/disconnection is reported over D-BUS in Linux, for ex, so that there is no need to constantly enumerate HID devices to see whether something new got connected. Together with supporting something like hidapi so that the code can be shared between Windows/Linux and Mac and doesn't need to be written to use the low-level platform-specific APIs. And that would be a really major change to Wiiuse ... 

That's why it was originally proposed as a "Wiiuse 2.0" project, it just doesn't fit well into the current setup. It might be possible to just hack in HID support on Linux without all that extra functionality but then the code would more-or-less behave like the Windows one - you would need to use the OS tools to connect to the Wiimote first and then start a Wiiuse app or it won't find and connect to the device :(

**Comment by Squall-Leonhart** [comment]
HIDAPI isn't compatible with the Dolphinbar, as it presents a virtualized HID interface instead of the Raw/Native one of the Wiimotes themselves.

**Comment by janoc** [comment]
Thanks for the info. However, which platform are you talking about? What do you mean by "virtualized HID interface"? 

I have used Wiiuse (i.e. the raw HID interface) with DolphinBar in Windows in the past without issues. DolphinBar has multiple modes and one of them is basically a pass-through of the HID communication with the Wiimote over USB to the host. That one works with Wiiuse without issues in Windows because it presents itself as a normal HID device, the same as if you connect the Wiimote directly to the PC via Bluetooth first. 

I haven't tried to use it with Linux but I would expect it to show up as a standard /dev/input/* device and /dev/hidraw* device on connection, as any other USB HID device. 

I don't see why it wouldn't work with HIDAPI but haven't tested it.
